### DISCUSSION TITLE
ProdEnvMatA

### DISCUSSION THREAD

**Discussion by npiroozan** [deepmodeling/deepmd-kit#1714] [Q&A]
Good Afternoon,

I hope you are doing well.
<img width="775" alt="MicrosoftTeams-image" src="https://user-images.githubusercontent.com/87953329/168887955-75c3fba9-a188-4c71-a442-c38e11438d87.png">

I noticed something interesting in the Tensorboard trace using se_e2_a as the descriptor.  After tuning OMP, TF_Intra_OP, and TF_Inter_OP Parallelism a remaining bottleneck is ProdEnvMatA.  Attached please find the respective tensorboard trace.

This was run on a CPU node with 8 workers and a local batch size of 16.  As one can see, all tf_computes are held up initially until ProdEnvMatA is complete.  May I please ask 2 questions regarding this op:

1.  What is the function of this Op.
2.   Is there any way to reduce its compute time?

Thank you very much!

**Comment by wanghan-iapcm** [comment]
@denghuilu Please take a look a the question. Thanks!

**Comment by njzjz** [comment]
This Op computes the environmental matrix. See eq. 5 in [this page](https://tutorials.deepmodeling.com/en/latest/Tutorials/DeePMD-kit/learnDoc/Theory.html).

I strongly agree that it can be further optimized.

**Reply by npiroozan** [reply]
I see, thank you.  Perhaps it can be parallelized in the same manner as ProdForceSeA?
